### PR TITLE
SNOW-1369280 Close channels asynchronously

### PR DESCRIPTION
**Pull request by sfc-gh-lshcharbaty**
# Overview

SNOW-1369280
By closing channels in parallel, speeds up task rebalancing. As in  #839, `SFException` is ignored and doesn't fail a connector task.

## Pre-review checklist
- [ ] This change should be part of a Behavior Change Release. See [go/behavior-change](http://go/behavior-change).
- [ ] This change has passed Merge gate tests
- [ ] Snowpipe Changes
- [x] Snowpipe Streaming Changes
- [ ] This change is TEST-ONLY
- [ ] This change is README/Javadocs only
- [x] This change is protected by a config parameter `snowflake.streaming.closeChannelsInParallel.enabled`.
    - [ ] `Yes` - Added end to end and Unit Tests. 
    - [ ] `No` - Suggest why it is not param protected
- [ ] Is his change protected by parameter <PARAMETER_NAME> on the server side?
    - [ ] The parameter/feature is not yet active in production (partial rollout or PrPr, see [Changes for Unreleased Features and Fixes](http://go/ppp-prpr)).
    - [ ] If there is an issue, it can be safely mitigated by turning the parameter off. This is also verified by a test (See [go/ppp](http://go/ppp)).

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 1]
```diff
@@ -349,15 +361,40 @@ public void callAllGetOffset() {
...
-    partitionsToChannel.clear();
+  }
 
-    StreamingClientProvider.getStreamingClientProviderInstance()
-        .closeClient(this.connectorConfig, this.streamingIngestClient);
+  private void closeAllInParallel() {
+    CompletableFuture<?>[] futures =
+        partitionsToChannel.entrySet().stream()
+            .map(
+                entry -> {
```
Comment (sfc-gh-gjachimko) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:388]: i think i'd combine the two variants of the close methods together,
just replace the return with
if  (closeChannelsInParallel) {
return topicPartitionChannel.closeChannelAsync() ;
} else {
   topicPartitionChannel.closeChannel();
 return CompletableFuture.completedFuture();
}

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java [hunk 2]
```diff
@@ -125,6 +125,11 @@ public class SnowflakeSinkConnectorConfig {
   public static final boolean SNOWPIPE_FILE_CLEANER_FIX_ENABLED_DEFAULT = false;
   public static final int SNOWPIPE_FILE_CLEANER_THREADS_DEFAULT = 1;
 
+  // Whether to close streaming channels in parallel.
+  public static final String SNOWPIPE_STREAMING_CLOSE_CHANNELS_IN_PARALLEL =
+      "snowflake.streaming.closeChannelsInParallel.enabled";
+  public static final boolean SNOWPIPE_STREAMING_CLOSE_CHANNELS_IN_PARALLEL_DEFAULT = false;
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:131]: This doesn't feel like a configuration should be exposed to customer. As a customer, I don't know whether I should close the channels in parallel or not. In fact, I don't really care, I just want my system to run without any failure. WDYT?

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:131]: agreed but I think we want to try this out for customers who has a lot of partitions in their single task. so enabling by default will not benefit a lot of customers 
- also its good to have param protection. 
- We should stress test this and slow roll out in couple of releases.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 3]
```diff
@@ -392,11 +443,33 @@ public void close(Collection<TopicPartition> partitions) {
...
-        partitionsToChannel.size());
+        "Closing partitionChannel:{}, partition:{}, topic:{}",
+        topicPartitionChannel == null ? null : topicPartitionChannel.getChannelNameFormatV1(),
+        topicPartition.topic(),
+        topicPartition.partition());
+
+    // It's possible that some partitions can be unassigned before their respective channels are
+    // even created.
+    return topicPartitionChannel == null
+        ? CompletableFuture.completedFuture(null)
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:469]: thinking out loud, does this need a NPE handling? or we are fine in `closeInParallel`?

File: test/rest_request_template/schema_evolution_nullable_values_after_smt.json [hunk 4]
```diff
@@ -15,6 +15,7 @@
     "snowflake.schema.name": "SNOWFLAKE_SCHEMA",
     "snowflake.role.name": "SNOWFLAKE_ROLE",
     "snowflake.ingestion.method": "SNOWPIPE_STREAMING",
+    "snowflake.streaming.closeChannelsInParallel.enabled": true,
```
Comment (sfc-gh-japatel) [inline test/rest_request_template/schema_evolution_nullable_values_after_smt.json:18]: can we add a test with false to make sure rollback is working like current behavior?

**Review by sfc-gh-japatel — COMMENTED** [review]
added couple of comments but overall LGTM!

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java (hunk 2 above)
Comment (sfc-gh-xhuang) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:131]: What's the risk of having it default true and then getting rid of it in a future release?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java (hunk 1 above)
Comment (sfc-gh-lshcharbaty) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:388]: I want to keep the existing flow intact. Yes that introduces some duplication but I know for sure that nothing is changed when the parameter is disabled.

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java (hunk 2 above)
Comment (sfc-gh-lshcharbaty) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:131]: @sfc-gh-tzhang 
This is not a publicly announced parameter. It's going to behave as a knob that's going to allow tho customer to easily roll the change back in case something goes wrong. Exactly what @sfc-gh-japatel wrote.

@sfc-gh-xhuang 
Imo it's better to be safe than sorry.  I'm not sure how the parallelization is going to work on the customer environments as channel closing is assigned to a shared `ForkJoinPool` that seems to exist as a single instance and shared across all tasks on the same host.

I suggest we approach the "change rollout" parameters the following way:
1. Protect the change with a parameter and disable it by default. If some customers experience problems that can be solved by the change, they can turn it on. 
1. Enable the change by default. If some customers experience problems that cause by the change, they can turn it off and communicate to us.
1. Get rid of the older implementation.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 5]
```diff
@@ -349,15 +361,40 @@ public void callAllGetOffset() {
...
+                entry -> {
+                  String channelKey = entry.getKey();
+                  TopicPartitionChannel topicPartitionChannel = entry.getValue();
+
+                  LOGGER.info("Closing partition channel:{}", channelKey);
+                  return topicPartitionChannel.closeChannelAsync();
+                })
+            .toArray(CompletableFuture[]::new);
+
+    CompletableFuture.allOf(futures).join();
```
Comment (sfc-gh-wtrefon) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:397]: can't we reuse here the closeInParallel(Collection<TopicPartition> partitions) and pass to it `partitionsToChannels`

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java (hunk 3 above)
Comment (sfc-gh-lshcharbaty) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:469]: No, null handling isn't needed. 
1. The returned future carries a `Void` value, so only `null` can be passed in there.
2. We don't do anything with the values returned by futures. The only thing we care about is their completion.

File: test/rest_request_template/schema_evolution_nullable_values_after_smt.json (hunk 4 above)
Comment (sfc-gh-lshcharbaty) [reply test/rest_request_template/schema_evolution_nullable_values_after_smt.json:18]: I'll try to test this in a unit test.
Each e2e test is run about 45 times, so I'd like to avoid writing e2e's for functionality that can be tested within a quicker set of tests.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java (hunk 3 above)
Comment (sfc-gh-gjachimko) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:469]: perhaps these futures could return "key" instead - then you could simply pass method pointer to the thenAccept(partitonsToChannel::remove) ?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 6]
```diff
@@ -1098,6 +1103,59 @@ public void closeChannel() {
...
+        .exceptionally(this::tryRecoverFromCloseChannelError);
+  }
+
+  private CompletableFuture<Void> closeChannelWrapped() {
+    try {
+      return this.channel.close();
+    } catch (SFException e) {
+      // Calling channel.close() can throw an SFException instead of putting it into a completable
+      // future.
+      CompletableFuture<Void> future = new CompletableFuture<>();
```
Comment (sfc-gh-gjachimko) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1122]: this comment is a bit odd - it took me a few reads to understand your point. 
i'd either get rid of it completely - i think it is clear that the sync part of the channel.close can throw or rephrase it to something like "wrap SFException into CompletableFuture to keep consistent method chain"

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 7]
```diff
@@ -1098,6 +1103,59 @@ public void closeChannel() {
...
+      future.completeExceptionally(e);
+      return future;
```
Comment (sfc-gh-dseweryn) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1124]: A bit shorter
```suggestion
      return CompletableFuture.failedFuture(e);
```

Comment (sfc-gh-lshcharbaty) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1124]: `CompletableFuture.failedFuture` was introduced in Java 9 only :(

File: test/rest_request_template/schema_evolution_nullable_values_after_smt.json (hunk 4 above)
Comment (sfc-gh-lshcharbaty) [reply test/rest_request_template/schema_evolution_nullable_values_after_smt.json:18]: Added [CloseTopicPartitionChannelIT](https://github.com/snowflakedb/snowflake-kafka-connector/pull/841/files#diff-24f62287e02cb6967f8a0c0792c4bab2ef5fcec5460003c41b951daf399dc6f1R17)  that ensures all channels are closed no matter what the config value is.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java (hunk 3 above)
Comment (sfc-gh-lshcharbaty) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:469]: `TopicPartitionChannel` which returns a future from `closeChannelAsync` knows nothing about a "key"`. Therefore, it can't return it.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 7 above)
Comment (sfc-gh-dseweryn) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1124]: Oh :C

**Review by sfc-gh-dseweryn — APPROVED** [review]
LGTM

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java (hunk 2 above)
Comment (sfc-gh-xhuang) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:131]: Understood but I was just thinking that there is not much difference/value between steps 1 and 2 but yes we can be a bit safer this way.

Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:131]: Would this still be a behavior change after we finish step 3? This configuration is gone and customer's KC won't be able to start, then they will need to get the logs to understand why. Assuming they know how to look into the log, then they realize that the configuration is being removed and they will probably need to reach out to Snowflake to understand why it's being removed?

It seems to me a better way is to do a beta release with the fix so customer can try it (or just provide a private jar)? Other customers won't use this version since it's a beta version.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 8]
```diff
@@ -349,15 +361,40 @@ public void callAllGetOffset() {
...
         (partitionChannelKey, topicPartitionChannel) -> {
           LOGGER.info("Closing partition channel:{}", partitionChannelKey);
           topicPartitionChannel.closeChannel();
         });
-    partitionsToChannel.clear();
+  }
 
-    StreamingClientProvider.getStreamingClientProviderInstance()
-        .closeClient(this.connectorConfig, this.streamingIngestClient);
+  private void closeAllInParallel() {
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:384]: Think about more I'm not sure this approach would help a lot. If you look into the close function in the SDK, it will flush everything in the client and then call get status on this particular channel. If we're doing it in parallel, the same number of flush will be called, every flush will still be run in serial and block other flushes. I thought most of time is spent on waiting for server side to commit, which will stay the same since everything will be flush at once. Have you verified locally that it reduces the close time dramatically with a large number of channels?

Comment (sfc-gh-akowalczyk) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:384]: Hi @sfc-gh-tzhang, as Lex is OOO, I prepared a short test for that. I am not sure whether the IT is sufficient to give meaningful information, but I ran [such a test](https://github.com/snowflakedb/snowflake-kafka-connector/commit/7d6a84236a7ba7522e3a713e2fe5e61a151f20fb#diff-4005e4dba838ef1cd8e0fe2c4b34e061ed69ec1da322bd1ebf1503e8be051a1eR225-R269) closing 50 channels. It shows the parallelism speeded the execution up.

13513.7 ms mean of sequential execution time vs 1508.0 of parallel
image

But the scenario does not cover the case where we have rows present in `net.snowflake.ingest.streaming.internal.SnowflakeStreamingIngestChannelInternal` buffer (the scenario described by you) - I cannot reproduce it even though I played with the buffering parameters.

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java (hunk 2 above)
Comment (sfc-gh-akowalczyk) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:131]: @sfc-gh-tzhang I don't see any schema validation for configuration properties. We will just stop respecting the `snowflake.streaming.closeChannelsInParallel.enabled` config key, but no errors should be observed. So, I suppose there will be no behavioral change afterwards.

**Review by sfc-gh-tzhang — APPROVED** [review]
Approved to unblock the PR, but I think this change is a risky change since I'm not sure if something weird would happen if we close a few hundreds of channels in parallel, since every close will spin up a background flush thread. Also another concern is calling get status in parallel would potentially overload DS node. I think longer term, the better solution is to add a close function which doesn't call flush or get status so it will be very fast. Could we create a JIRA to track this? Thanks!

**Comment by ebuildy** [comment]
As this is for better performance, is it a way to monitor it please?

This could help debugging crash/rebalance issues.

Thanks you, cant wait to test it!

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java (hunk 8 above)
Comment (sfc-gh-akowalczyk) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:384]: The log level was set to DEBUG so it added a few additional seconds to the measured time. I changed it to INFO and attached the profiler. It takes ~11 seconds and spends most of its time on `CompletableFuture.get()`. 
image

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 9]
```diff
@@ -349,15 +361,40 @@ public void callAllGetOffset() {
...
+                  String channelKey = entry.getKey();
+                  TopicPartitionChannel topicPartitionChannel = entry.getValue();
+
+                  LOGGER.info("Closing partition channel:{}", channelKey);
+                  return topicPartitionChannel.closeChannelAsync();
+                })
+            .toArray(CompletableFuture[]::new);
+
+    CompletableFuture.allOf(futures).join();
   }
```
Comment (ebuildy) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:398]: In my small comprehension,  `CompletableFuture` uses the `ForkJoinPool.commonPool()` as its default thread pool with `(Runtime.getRuntime().availableProcessors() - 1)` threads. We can cheat by setting JVM cpu count but I rather suggest to create a thread pools like this:

```
import java.util.concurrent.*;

private void closeAllInParallel() {
    ExecutorService executor = Executors.newFixedThreadPool(CONFIG_THREAD_POOL_SIZE);
    CompletableFuture<?>[] futures =
        partitionsToChannel.entrySet().stream()
            .map(
                entry -> {
                  String channelKey = entry.getKey();
                  TopicPartitionChannel topicPartitionChannel = entry.getValue();

                  LOGGER.info("Closing partition channel:{}", channelKey);
                  return CompletableFuture.runAsync(topicPartitionChannel::closeChannelAsync, executor);
                })
            .toArray(CompletableFuture[]::new);

    CompletableFuture.allOf(futures).join();
    executor.shutdown(); // Don't forget to shutdown the executor
}
```

**Comment by ebuildy** [comment]
> Approved to unblock the PR, but I think this change is a risky change since I'm not sure if something weird would happen if we close a few hundreds of channels in parallel, since every close will spin up a background flush thread. Also another concern is calling get status in parallel would potentially overload DS node. I think longer term, the better solution is to add a close function in the SDK which doesn't call flush or get status so it will be very fast. Could we create a JIRA to track this? Thanks!

The default thread pool used by `CompletableFuture`  according doc at https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/CompletableFuture.html 

```
All async methods without an explicit Executor argument are performed using the [ForkJoinPool.commonPool()](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/ForkJoinPool.html#commonPool())
``` 

Would be better to create a static thread pool

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java (hunk 9 above)
Comment (sfc-gh-akowalczyk) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:398]: Thanks, @ebuildy, for providing an example. Let me merge the PR in its current form, I'll check with the SDK team on a possible built-in solution. If the timeline is not optimistic, I'll propose this change to the rest of the team.

**Comment by sfc-gh-akowalczyk** [comment]
> Approved to unblock the PR, but I think this change is a risky change since I'm not sure if something weird would happen if we close a few hundreds of channels in parallel, since every close will spin up a background flush thread. Also another concern is calling get status in parallel would potentially overload DS node. I think longer term, the better solution is to add a close function in the SDK which doesn't call flush or get status so it will be very fast. Could we create a JIRA to track this? Thanks!

Thanks, created SNOW-1437461

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java (hunk 9 above)
Comment (ebuildy) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:398]: fantastic! Any plan to do the same for open channel operations?
